InfoBar Control (#94)
microsoft/microsoft-ui-xaml-specs: active/InformationBar/InformationBar.md

Due to their visual layouts, inherent intrusiveness, or available features they are not sufficient for displaying notifications at an app-wide level.

This spec introduces an InfoBar WinUI (Xaml) control that an app can use for these kinds of messages.

# InfoBar Class

Represents an informational control that is visible but does not interfere with the user. 

## Is this the right control?
Use an InfoBar control when a user should be informed of, acknowledge, or take action on a message. By default the notification will remain in the content area until dismissed by the user but will not necessarily break user flow.

An InfoBar will take up space in your layout and behave like any other child elements. It will not cover up other content or float on top of it.

Do not use an InfoBar control to confirm or respond directly to a user action, for time-sensitive alerts, or for non-essential messages.

## Remarks
Use an InfoBar that is dismissed by the user or when the status is resolved for scenarios that **directly** impact app perception or experience

> - Internet connectivity lost

Reviewer: nit: list seems to be disconnected from the sentence above. Perhaps add a line like. 'Here are some examples'

- For scenarios where a persistent notification is not needed, e.g. displaying information in context of a specific UI element, a [Flyout](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/flyouts) is a better option. 
- For scenarios where the application is confirming a user action, showing information the user ***must*** read, use a [ContentDialog](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/dialogs).
  - Additionally, if a status change to the app is so severe that it needs to block all further ability for the user to interact with the app, use a ContentDialog.
- For scenarios where a notification is a transient teaching moment, a [TeachingTip](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/teaching-tip) is a better option.

For more info about choosing the right notification control, see the [Dialogs and Flyouts](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/) article.

# Examples

## Create an InfoBar

> The XAML below describes an inline InfoBar with the default styling for an informational notification. An info bar can be created anywhere in the element tree or code behind. In this example the InfoBar expands to fill the width of the StackPanel layout control it resides in.

Reviewer: >In this example the InfoBar expands to fill the width of the StackPanel layout control it resides in. [](start = 175, length = 101)

I don't see the StackPanel in the example below. I don't think we need this since it behaves like any other control.

# Background
> This spec corresponds to [issue 913](https://github.com/microsoft/microsoft-ui-xaml/issues/913) on the WinUI repo.

Users should be informed about essential status changes that occur on an app level.
These status changes affect the app as a whole and can be either critical or informational.
Critical status changes like lost internet connectivity are directly impactful to app functionality while informational status changes like an update has completed and been applied are indirectly impactful to app functionality.
These notifications and corresponding information should be presented in a consistent, predictable, and relevant way to the user depending on the specific scenario.

> Currently, Teaching Tip, Content Dialog, and customizations of other flyouts and dialogs exist as options to show these notifications but these controls were not specifically designed to handle app-wide status change notifications. 

Reviewer 2: About "other ... dialogs", what other dialogs beyond ContentDialog?

Reviewer 3: There's also the MessageDialog.

Reviewer 2: Good point, thanks.

## Is this the right control?
Use an InfoBar control when a user should be informed of, acknowledge, or take action on a message. By default the notification will remain in the content area until dismissed by the user but will not necessarily break user flow.

An InfoBar will take up space in your layout and behave like any other child elements. It will not cover up other content or float on top of it.

Do not use an InfoBar control to confirm or respond directly to a user action, for time-sensitive alerts, or for non-essential messages.

### Remarks
Use an InfoBar that dismiss via the user or when the status is resolved for scenarios that **directly** impact app perception or experience ⚠

- Internet connectivity lost
- Error while saving a document when triggered automatically, not related to specific user action
- No microphone plugged in when attempting to record
- Can't connect to your phone
- The subscription to the application is expired

Use an InfoBar that dismiss via the user or a timer for scenarios that **indirectly** impact app perception or experience ℹ

- A call has begun recording
- Update applied with link to 'Release Notes'

> - The terms of service have been updated and require acknowledgement

Reviewer 2: If it requires user acknowledgement, wouldn't it be bad for a timer to dismiss it?

- For scenarios where a persistent notification is not needed, e.g. displaying information in context of a specific UI element, a [Flyout](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/flyouts) is a better option. 
- For scenarios where the application is confirming a user action, showing information the user ***must*** read, use a [ContentDialog](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/dialogs).
  - Additionally, if a status change to the app is so severe that it needs to block all further ability for the user to interact with the app, use a ContentDialog.
- For scenarios where a notification is a transient teaching moment, a [TeachingTip](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/teaching-tip) is a better option.

For more info about choosing the right notification control, see the [Dialogs and Flyouts](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/) article.

# Examples

An info bar can have several configurations, here are some notable ones.

> When an InfoBar is conveying information of a common severity, like displaying an error or warning, it can be set to have one of many severity levels to use consistent [Fluent](https://docs.microsoft.com/en-us/windows/apps/fluent-design-system) styling for it's identifiers.

Reviewer 2: I don't understand the connection between the text and the picture. Is this example using a severity level? I would just remove this first section, up to the "Create an InfoBar" example, as it's redundant of the later examples. It provides an introduction/overview of the follow examples, but that's not a style that we use in the docs. It's better to jump right in to the first example.

> When an InfoBar is conveying information of a common severity, like displaying an error or warning, it can be set to have one of many severity levels to use consistent [Fluent](https://docs.microsoft.com/en-us/windows/apps/fluent-design-system) styling for it's identifiers.

Reviewer 2: These examples should include example code

By default, the InfoBar will not be visible. Set the IsOpen property to true in the XAML or code behind to display the information.

XAML
```xml
<Grid>
    <!--Other UI content -->
    <StackPanel x:Name="ContentArea">
        <controls:InfoBar x:Name="UnsuccessfulSaveNotification"
            Severity="Warning"
            Title="Error while saving"
            Message="Your document was unable to be saved.">
        </controls:InfoBar>
    </StackPanel>
    <!--Other UI content -->
<Grid>
```
C#
```C#
public MainPage()
{
    this.InitializeComponent();

    if(!SaveSuccessful())
    {
        UnsuccessfulSaveNotification.IsOpen = true;
    }
}
```

Here is the visual representation of the info bar in the page.

![A mockup of an InfoBar in a Warning state with a close button and a message](images/Warning_NoButton.png)

## Using pre-defined severity levels
The type of the info bar can be set via the Severity property to automatically set a consistent status color, icon, and assistive technology settings dependent on the criticality of the notification.

> Mockups in various Severities to be added.

Reviewer 2: We don't usually work with mockups in the documentation. These would be better as examples (including the markup that creates them).

Author: When you mean mockups do you mean the File Explorer sample app screenshots? I can create a simpler sample app to grab screenshots to highlight the functionality if that's valuable at this stage. 
For the content wrapping example, showing just the InfoBar on its own is not as valuable as showing where it is on the page and its relation to other content. For the markup snippet, is it beneficial to include the accompanying layout controls?

![Mockup of InfoBar with no content in different Severity colors and icons for light mode](images/Docked_ToolkitSeverity.png)

![Mockup of InfoBar with no content in different Severity colors and icons for dark mode](images/Docked_ToolkitSeverityDark.png)

## Programmatic dismiss in info bar

> An info bar can be dismissed by the user via the close button or programmatically. If the notification is required to be in view until the status is resolved and you would like to remove the ability for the user to dismiss the info bar, you can set the IsUserDismissable property to false.

Reviewer 2: "Close" is a more standard term for this action than "dismiss". And this looks more like a case for the "Can" pattern. So instead of "IsUserDismissable" I think "CanClose" would be a better name. (For example similar to [TextBox.CanUndo](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.TextBox.CanUndo).)

Author: I'm open to that, could we leave this as something to be discussed at the API Review?

By default, the close button will appear as an 'X' and the IsUserDismissable property is therefore set to true.

XAML
```xml
<StackPanel x:Name="ContentArea">
    <controls:InfoBar x:Name="NoInternetNotification"
        Severity="Error"
        Title="No Internet"
        Message="Reconnect to continue working."
        IsUserDismissable="False">
    </controls:InfoBar>
</StackPanel>
```
![Mockup of an InfoBar in a Critical state with no close button](images/Critical_IconNoClose.png)

## Custom styling: status color and icon
Outside of the pre-defined severity levels, the Background and IconSource properties can be set to customize the icon and background color. The info bar will retain the assistive technology settings of the severity defined, or default if none was defined.

> A custom background color can be set via the standard Background property and will override the color set by Severity. Please keep in mind content readability and accessibility when setting your own color.

Author: @MikeHillberg, @ranjeshj, @teaP 

Wanted to highlight an implementation distinction here. Currently, in contrast to the guidance here, the background color of the InfoBar can **not** be custom set via the Background property. Should it? Or, because background color is tied to the Severity level, that we recommend users to overwrite the associated Severity level background brush? @teaP mentioned that if we wanted to allow the Background property to override the background color set by the Severity that we could implement a background panel as an in-between for the Severity brushes and Background property.

Thoughts?

Reviewer 4: Should the severity set a base value which gets overwritten by the Background and Foreground chosen by the app devs?

Or does Severity override all?

Is there any contextual or functional value to having a set severity, but have the look of the control set differently by the app?

Author: Conceptually, a Severity is always going to be "set" in that even if a developer doesn't explicitly define a Severity it will follow the "informational" design with a pre-defined background color, icon, and accessibility settings. With that, I don't think Severity should override other custom values and should instead be a baseline.

My opinion is that if a developer wants to overwrite the background color independent of Severity they can. We're allowing them to set a custom icon without having them change the definition of the Severity for their app so why should there be a different pattern for the Background color?

Reviewer 2: It's unfortunate that Background is on Control, since it doesn't necessarily make sense on every control. But in this case I'd expect it to override whatever the control would otherwise do, regardless of the information level.

Reviewer 5: No problem, I'll make it work. 👍

    <controls:InfoBar x:Name="UnsuccessfulSaveNotification"
        Severity="Warning"
        Title="Error while saving"
        Message="Your document was unable to be saved.">
        <controls:InfoBar.ActionButton>
            <Button Content="Try Again" Command="{x:Bind SaveDocuments}"/>
        </controls:InfoBar.ActionButton>
    </controls:InfoBar>
</StackPanel>
```

![A mockup of an InfoBar with a single line message and an action button](images/Warning_Button.png)

XAML
```xml
<StackPanel x:Name="ContentArea">
    <controls:InfoBar x:Name="ConnectionErrorNotification"
        Title="Error while saving"
        Message="Lorem ipsum long message">
        <controls:InfoBar.ActionButton>
            <HyperlinkButton
                Content="www.microsoft.com" 
                NavigateUri="http://www.microsoft.com"/>
        <controls:InfoBar.ActionButton/>
    </controls:InfoBar>
</StackPanel>
```
![A mockup of an InfoBar with a single line message and hyperlink](images/Default_MessageHyperlink.png)
![A mockup of an InfoBar with a message expanding multiple lines and a hyperlink](images/Docked_MultiLineIconTitleHyperlink.png)

## Custom content

> Content can be added to an InfoBar using the Content property. It will appear below the message and above the action or hyperlink buttons if defined and multiline. The InfoBar will expand to fit the content defined.

Reviewer 2: What does it mean to be multiline?

(Maybe this is answered in the next example and they just need to re-ordered)

        Title="Message Title"  
        Message="This message is very long, so long in fact it needs to wrap to a second line in the notification">
    </controls:InfoBar>
</StackPanel>
```

![A mockup of a multiline message InfoBar](images/Docked_MultiLineIconTitle.png)
![A mockup of a multiline title and message InfoBar](images/Docked_MultiLineLongTitle.png)

## Canceling and deferring close
The Closing event can be used to cancel and/or defer the close of an InfoBar. This can be used to keep the InfoBar open or allow time for an action or custom animation to occur. When the closing of an InfoBar is canceled, IsOpen will go back to true, however, it will stay false during the deferral. A programmatic close can also be canceled.

XAML
```xml
<controls:InfoBar x:Name="UpdateAvailable"
    Title="Update Available"
    Message="Please close this tip to apply required security updates to this application"
    Closing="InfoBar_Closing">
</controls:InfoBar>
```
C#
```C#
public void InfoBar_Closing(InfoBar sender, InfoBarClosingEventArgs args)
{
    if (args.Reason == InfoBarCloseReason.CloseButton) {
        // if scenario failed
        args.Cancel = true;
    }
    
}
```

> # Inputs and Accessibility

Reviewer 2: The public docs don't usually go into this much detail, but it's great detail. This can be moved to an Appendix of this spec.

| Severity |  NotificationProcessing | Behavior in Narrator|
|:--- | :---| :---|
| Error | NotificationProcessing_ImportantAll| Add new item to the end of the queue. It doesn’t matter the source, what is currently speaking or any other items in the queue queued with a higher priority than normal text. <br><br> This will cause it to get spoken after the current utterance/string finishes. Meaning it won't interrupt the speech but it will queue itself next to be spoken, ahead of anything already queued. <br><br> Focus and keyboard will NOT interrupt or flush, pressing control will silence and flush all |
| Warning | NotificationProcessing_ImportantAll| See above |
| Success | Processing_All| Add new item to the end of the queue. <br><br> Will be added to the end of the queue meaning all existing queued text will need to be spoken before this text will be spoken. <br><br>Focus, keyboard and control will silence/flush them all
| Default | Processing_All | See above

> #### Narrator

Reviewer 2: Why are we talking about Narrator? Are we making updates to that app, or does this section apply to any screen reader?

Author: We are not making updates to Narrator however when describing the effect of the [UIA Event](https://docs.microsoft.com/en-us/windows/win32/winauto/uiauto-eventsoverview) to screen readers I've seen Narrator used in other specs as example behavior.

 - Error: An error or problem that has occurred.
 - Warning: A condition that might cause a problem in the future.
 - Success: A long-running and/or background task has completed.
 - Default: General information that requires the user's attention.

Icons and color should not be the only UI components signifying meaning for your notification. Text in the notification's Title and/or Message should be included to display information.

## Message 

Text in your notification will not be a constant length in all languages. For the Title and Message property this may impact whether your notification will expand to a second line. 
We recommend you avoid positioning based on message length or other UI elements set to a specific language.

The notification will follow standard mirroring behavior when localized to/from languages that are right to left (RTL) or left to right (LTR). The icon will only mirror if there is directionality.

Please view the guidance for [Adjust layout and fonts, and support RTL](https://docs.microsoft.com/en-us/windows/uwp/design/globalizing/adjust-layout-and-fonts--and-support-rtl) for more information about text localization in your notification.

> # Remarks

Reviewer 2: How are these remarks different than the earlier remarks? Where are there two versions?

 - Example: Terms and Conditions of the application have been updated and need to be acknowledged for continued use
   - A default info bar could appear notifying the user the T&C have updated and link to the contents. The user acknowledges these updated T&C through dismissing the notification.
 - Example: A virus scan that was running in the background of your application has completed
   - A success-styled info bar could appear notifying the user the scan has completed and direct them to view the results.

 Note: Users should **always** have the option to dismiss informational notifications.

# API Notes

### Notable Properties  

| Name | Description |
|:-:|:--|
| IsOpen| Gets or sets a value that determines the visibility of the InfoBar. By default, is set to false. |
| Severity | Gets or sets a value that indicates the  color and icon to style the InfoBar. It will also define the assistive technology settings. |
| IsUserDismissable| Gets or sets a boolean that indicates whether the user will be able to dismiss the InfoBar.

### Events  
| Name | Description |
|:-:|:--|

> | CloseButtonClick | Occurs after the close button has been tapped. |

Reviewer 2: "tapped" => "clicked"  ("Tap" is a touch term, and you can use a mouse or keyboard on a button.)

# Background
> This spec corresponds to [issue 913](https://github.com/microsoft/microsoft-ui-xaml/issues/913) on the WinUI repo.

> Users should be informed about essential status changes that occur on an app level.

Reviewer 2: Would it be easy to add a couple of pictures?

Author: Do you mean throughout the document or in the Background section specifically? There should be pictures/mockups throughout the spec...

These status changes affect the app as a whole and can be either critical or informational.
Critical status changes like lost internet connectivity are directly impactful to app functionality while informational status changes like an update has completed and been applied are indirectly impactful to app functionality.
These notifications and corresponding information should be presented in a consistent, predictable, and relevant way to the user depending on the specific scenario.

Currently, Teaching Tip, Content Dialog, and customizations of other flyouts and dialogs exist as options to show these notifications but these controls were not specifically designed to handle app-wide status change notifications. 
Due to their visual layouts, inherent intrusiveness, or available features they are not sufficient for displaying notifications at an app-wide level.

> 

Reviewer 2: ```suggestion

This spec introduces an InfoBar API to ...
```

# Description

> An InfoBar is a persistent, actionable, app-wide notification intended for displaying critical and/or actionable information that impact app perception or user experience.

Reviewer 2: (Minor comment at this stage) This text is the IntelliSense you'll get from typing <InfoBar>. For Xaml controls it always seems to start with "Represents", and tends to focus more on the 'what'.

```suggestion
Represents an informational control that is visible but does not interfere with the user. 
```

## Is this the right control?

> Use an InfoBar control when a user needs to be informed of, acknowledge, or take action on a message. By default the notification will remain in the content area until dismissed by the user but will not necessarily break user flow.

Reviewer 2: "User needs to take action" makes this sound modal, like a ContentDialog

An InfoBar will take up space in your layout and behave like any other child elements. It will not cover up other content or float on top of it.

> Do not use an InfoBar control to confirm or respond directly to a user action, for transient alerts, or for non-essential messages.

Reviewer 2: Why not use this for an essential transient message? Like network connectivity?

> ### Scenarios

Reviewer 2: ```suggestion
### Remarks
```

Use an InfoBar that dismiss via the user or when the status is resolved for scenarios that **directly** impact app perception or experience ⚠

- Internet connectivity lost
- Error while saving a document when triggered automatically, not related to specific user action
- No microphone plugged in when attempting to record
- Can't connect to your phone
- The subscription to the application is expired

Use an InfoBar that dismiss via the user or a timer for scenarios that **indirectly** impact app perception or experience ℹ

- A call has begun recording
- Update applied with link to 'Release Notes'
- The terms of service have been updated and require acknowledgement
- An app-wide backup has successfully, asynchronously completed
- The subscription to the application is close to expiring

> ### When should a different control be used?

Reviewer 2: ```suggestion
#### When should a different control be used?
```

> There are some scenarios where a Content Dialog, Flyout, or Teaching Tip may be more appropriate to use.

Reviewer 2: ```suggestion
There are some scenarios where a ContentDialog, Flyout, or TeachingTip may be more appropriate to use.
```

> - For scenarios where a persistent notification is not needed, i.e. displaying information in context of a specific UI element, a [Flyout](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/flyouts) is a better option. 

Reviewer 2: ```suggestion
- For scenarios where a persistent notification is not needed, e.g. displaying information in context of a specific UI element, a [Flyout](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/flyouts) is a better option. 
```

- For scenarios where the application is confirming a user action, showing information the user ***must*** read, use a [Content Dialog](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/dialogs).
  - Additionally, if a status change to the app is so severe that it needs to block all further ability for the user to interact with the app, use a Content Dialog.
- For scenarios where a notification is a transient teaching moment, a [Teaching Tip](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/teaching-tip) is a better option.

For more info about choosing the right notification control, see the [Dialogs and Flyouts](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/) article.

# Examples

An information bar can have several configurations, here are some notable ones.

> When an InfoBar is conveying information of a common severity, like displaying an error or warning, it can be set to have one of many severity levels to use consistent Fluent styling for it's identifiers.

Reviewer 2: Link to what "Fluent" means?

The information bar can also be customized with its optional properties and with XAML Content to include extra buttons and other UI elements.
![A mockup of an InfoBar with custom status color and content](images/Custom_IconColor.png)

## Create an InfoBar

The XAML below describes an inline InfoBar with the default styling for an error  notification. An information bar can be created anywhere in the element tree or code behind. In this example the InfoBar expands to fill the width of the StackPanel layout control it resides in.

By default, the InfoBar will not be visible. Set the IsOpen property to true in the XAML or code behind to display the information.

XAML
```xml
<Grid>
    <!--Other UI content -->
    <StackPanel x:Name="ContentArea">
        <controls:InfoBar x:Name="UnsuccessfulSaveNotification"
            Severity="Warning"
            Title="Error while saving"
            Message="Your document was unable to be saved.">
        </controls:InfoBar>
    </StackPanel>
    <!--Other UI content -->
<Grid>
```
C#
```C#
public MainPage()
{
    this.InitializeComponent();

    if(!SaveSuccessful())
    {
        UnsuccessfulSaveNotification.IsOpen = true;

>         SetHaveDisplayedErrorNotification();

Reviewer 2: What is this line for? Is it relevant to the example?

    }
}
```

Here is the visual representation of the information bar in the page.

![A mockup of an InfoBar in a Warning state with a close button and a message](images/Warning_NoButton.png)

> ## Using pre-defined severity levels

Reviewer 2: Add a code example?

> The type of the information bar can be set via the Severity property to automatically set a consistent status color, icon, and assistive technology settings dependent on the criticality of the notification.

Reviewer 2: ```suggestion
The type of the info bar can be set via the Severity property to automatically set a consistent status color, icon, and assistive technology settings dependent on the criticality of the notification.
```

## Programmatic dismiss in info bar
An info bar can be dismissed by the user via the close button or programmatically. If the notification is required to be in view until the status is resolved and you would like to remove the close button from view, you can set the IsCloseButtonVisible property to false.
By default, the close button will appear as an 'X' and the IsCloseButtonVisible property is therefore set to true.

XAML
```xml
<StackPanel x:Name="ContentArea">
    <controls:InfoBar x:Name="NoInternetNotification"
        Severity="Critical"
        Title="No Internet"
        Message="Reconnect to continue working."
        IsCloseButtonVisible="False">
    </controls:InfoBar>
</StackPanel>
```
![Mockup of an InfoBar in a Critical state with no close button](images/Critical_IconNoClose.png)

## Custom styling: status color and icon
Outside of the pre-defined severity levels, the Background and IconSource properties can be set to customize the icon and background color. The info bar will retain the assistive technology settings of the severity defined, or default if none was defined.

> A custom background color can be set via the standard Background property and will override the color set by Severity. Please keep in mind content readability and accessibility when setting your own color.

Reviewer 2: Are the accessibility features explained anywhere? Maybe should be in the remarks

> A custom icon can be set via the IconSource property left of the Title and Message in the InfoBar. By default, an icon will be visible. This icon can be removed by setting the IsIconVisible property to false. For custom icons, the recommended icon size is 20px.

Reviewer 2: Will it be scaled to 20px if necessary?

Author: I don't think they should as I don't believe other controls in the past have. @StephenLPeters is this correct for Teaching Tip?

> A custom icon can be set via the IconSource property left of the Title and Message in the InfoBar. By default, an icon will be visible. This icon can be removed by setting the IsIconVisible property to false. For custom icons, the recommended icon size is 20px.

Reviewer 2: ```suggestion
A custom icon can be set via the IconSource property. By default, an icon will be visible. This icon can be removed by setting the IsIconVisible property to false. For custom icons, the recommended icon size is 20px.
```

(Don't need to get into the current visual design details in the docs)

XAML
```xml
<StackPanel x:Name="ContentArea">
    <controls:InfoBar x:Name="NewPhotosNotification"
        Title="New Photos From Cloud"
        Message="Sync your latest photos to the Cloud"
        ActionButtonContent="Upload">

>         <controls:InfoBar.Background>

Reviewer 2: Can do both Background and IconSource as attributes rather than tags

            <SymbolIconSource Symbol="Camera" />
        </controls:InfoBar.IconSource>
    </controls:InfoBar>
</StackPanel>
```

![A mockup of an InfoBar in the top of the content area. The notification's title is "New Photos From Cloud" and it's message is "Sync your latest photos to the Cloud". It has a bright green accent color and a "Camera" icon.](images/Custom_IconColor.png)

## Add buttons
By default, an 'X' close button will appear as the right most component in the bar.

An additional action button can be added by setting the ActionButtonContent and ActionButtonCommand properties. There is also built-in support for a single hyperlink button to ensure text contrast remains accessible with the various background colors. We recommend that only a single action button or hyperlink button is set. Outside of these, additional action buttons and hyperlinks can be added via custom content.

XAML
```xml
<StackPanel x:Name="ContentArea">
    <controls:InfoBar x:Name="UnsuccessfulSaveNotification"
        Severity="Warning"
        Title="Error while saving"
        Message="Your document was unable to be saved."
        ActionButtonContent="Try again"

>         ActionButtonCommand="SaveDocuments">

Reviewer 2: ```suggestion
        ActionButtonCommand="{x:Bind SaveDocuments}">
```

    </controls:InfoBar>
</StackPanel>
```

![A mockup of an InfoBar with a single line message and an action button](images/Warning_Button.png)

XAML
```xml
<StackPanel x:Name="ContentArea">
    <controls:InfoBar x:Name="ConnectionErrorNotification"
        Title="Error while saving"
        Message="Lorem ipsum long message">

>         <controls:InfoBar.HyperlinkButtonContent>

Reviewer 2: It's odd that I can provide an actual HyperlinkButton object but not a Button object.

            <HyperlinkButton
                Content="www.microsoft.com" 
                NavigateUri="http://www.microsoft.com"/>
        <controls:InfoBar.HyperlinkButtonContent/>
    </controls:InfoBar>
</StackPanel>
```
![A mockup of an InfoBar with a single line message and hyperlink](images/Default_MessageHyperlink.png)
![A mockup of an InfoBar with a message expanding multiple lines and a hyperlink](images/Docked_MultiLineIconTitleHyperlink.png)

## Custom content
Content can be added to an InfoBar using the Content property. It will appear below the message and above the action or hyperlink buttons if defined. The InfoBar will expand to fit the content defined.

XAML

> ```xml

Reviewer 2: Looks like the wrong example code (hyperlink rather than content)

<StackPanel x:Name="ContentArea">
    <controls:InfoBar x:Name="TermsAndConditionsNotification"
        Title="Update Complete!"  
        Message="Before you can proceed you must read and accept the new Terms and Conditions.">
        <controls:InfoBar.HyperlinkButtonContent>
            <HyperlinkButton Content="Read here" NavigateUri="www.microsoft.com"/>
        </controls:InfoBar.HyperlinkButtonContent>
    </controls:InfoBar>
</StackPanel>
```

![A sketch of an InfoBar in its default state with a hyperlink and message](images/Information_CustomContent.jpg)
Mockup with custom content to be added.

## Content wrapping

> By default, the text set in the Message property will wrap vertically in the control underneath the other visual components.

Reviewer 2: Is there a way to override the default?

Author: Should there be? The alternative would be to allow the content to cut itself off once it is too long. That's an option that Fluent UI's [MessageBar](https://developer.microsoft.com/en-us/fluentui#/controls/web/messagebar) has in their isMultiline property but I struggle to see a scenario where we would want to support that.

Or are you more focused on the wording of "by default"?

Reviewer 2: Yeah, it says "by default" indicating that there's a non-default case, but not explaining it.

XAML
```xml

> <StackPanel x:Name="ContentArea">

Reviewer 2: When the info bar is closed, is there an animation? Or a recommended way for me to add an animation?

Author: Currently, no.  I think by default no animations should be shown for these inline InfoBars however if a developer wanted to include their own -- e.g. coming down from the closest orientation (top, bottom, left, right) -- then I think that should be supported. Via storyboards? @teaP?

Reviewer 2: The text in the Canceling example teases the possibility of an animation. Do we need to give an example of how to do one?

Author: Removed mention of animation 😊

  - Example: The user's subscription to your application has expired and many features are deactivated
    - An error information bar could remain in view with a potential action button to resubscribe.
    - Another option is to allow the user to dismiss a default-style InfoBar depending on how vital a subscription is to your application.

Recommended patterns for informational notifications where the user needs to view or react to essential information
 - Example: Terms and Conditions of the application have been updated and need to be acknowledged for continued use
   - A default information bar could appear notifying the user the T&C have updated and link to the contents. The user acknowledges these updated T&C through dismissing the notification.
 - Example: A virus scan that was running in the background of your application has completed
   - A success-styled information bar could appear notifying the user the scan has completed and direct them to view the results.

 Note: Users should **always** have the option to dismiss informational notifications.

# API Notes

### Notable Properties  

| Name | Description |
|:-:|:--|

> | IsOpen| Gets or sets a value that determines the visibility of the InfoBar. By default, is set to false. |

Reviewer 5: What is the reason to have this in addition to the usual Visibility property?

Reviewer 5: (I'm guessing, after further reading, that it's so there are events around Closing/Closed and the reason for it, etc.)

| Severity | Gets or sets a value that indicates the  color and icon to style the InfoBar. It will also define the assistive technology settings. |
| IsCloseButtonVisible| Gets or sets a boolean that indicates whether a close button will appear

### Events  
| Name | Description |
|:-:|:--|

> | ActionButtonClick | Occurs after the action button has been tapped. |

Reviewer 5: The comps sometimes show two action buttons, but it looks like there's only ever one, right?

sufficient for displaying notifications at an app-wide level.

This spec introduces an InfoBar WinUI (Xaml) control that an app can use for these kinds of messages.

# InfoBar Class
Represents an informational control that is visible but does not interfere with the user. 

## Is this the right control?
Use an InfoBar control when a user should be informed of, acknowledge, or take action on a message. 
By default the notification will remain in the content area until dismissed by the user 
but will not necessarily break user flow.

An InfoBar will take up space in your layout and behave like any other child elements. 
It will not cover up other content or float on top of it.

Do not use an InfoBar control to confirm or respond directly to a user action, for time-sensitive alerts, 
or for non-essential messages.

## Remarks
Use an InfoBar that is dismissed by the user or when the status is resolved for scenarios 
that **directly** impact app perception or experience

Here are some examples:
- Internet connectivity lost
- Error while saving a document when triggered automatically, not related to specific user action

> - No microphone plugged in when attempting to record

Reviewer 6: This sounds like it was responding to "user action" (start recording) but that is listed as a "do not use" above.

Author: I think the term "responding to user action" may be too ambiguous.
> Do not use an InfoBar control to confirm or react directly to a user action when the state of the app is unchanged, for time-sensitive alerts, or for non-essential messages.

The examples that @oldnewthing included all are related to a changed state of the app in that is is now recording and/or in a meeting. I would argue that the final example, "the meeting has ended" would not be a message shown by an InfoBar as the state of the app is back in its default mode.

Thoughts?

Reviewer 2: Is that a suggestion to change the guideline text?

Author: Yep.

Author: My initial comment a month ago wasn't clear, I added more text to clarify that the message should be when the app state changes.

a specific UI element, a 
[Flyout](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/flyouts) 
is a better option. 
- For scenarios where the application is confirming a user action, showing information the user ***must*** read, use a
[ContentDialog](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/dialogs).
  - Additionally, if a status change to the app is so severe that it needs to block all further ability for the user to interact with the app, use a ContentDialog.
- For scenarios where a notification is a transient teaching moment, a 
[TeachingTip](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/teaching-tip) 
is a better option.

For more info about choosing the right notification control, see the 
[Dialogs and Flyouts](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/dialogs-and-flyouts/)
article.

# Examples

## Create an InfoBar
The XAML below describes an inline InfoBar with the default styling for an informational notification. 

> An info bar can be created anywhere in the element tree or code behind and will expand 

Reviewer 6: "anywhere" in the tree -- does the parent need to support (multiple) children? E.g. could I put it on a `Border`, even if the `Border` already has content?

Reviewer: There is nothing special about InfoBar in this sense. It is just another control. Border supports only one child, so this can be the child of the border. If you need multiple children, you need to add some panel (like StackPanel or Grid) and then add InfoBar to that panel.

> to fill the width of the layout control it resides in.

Reviewer 2: It's a good thing to call out, not sure the best way to word it. The control will use the full layout slot provided by the parent, subject to other standard layout features, like setting the HorizontalAlignment.

</controls:InfoBar>
```
C#
```C#
public MainPage()
{
    this.InitializeComponent();

    if(IsUpdateAvailable())
    {
        UpdateAvailableNotification.IsOpen = true;
    }
}
```

![A mockup of an InfoBar in the default state with a close button and a message](images/Default_TitleMessage.png)

## Using pre-defined severity levels
The type of the info bar can be set via the Severity property to automatically set a consistent status color, icon, and assistive technology settings dependent on the criticality of the notification. If no Severity is set, the default informational styling is applied.

XAML
```xml
<controls:InfoBar x:Name="SubscriptionExpiringNotification"
    Severity="Warning"
    Title="Your subscription is expiring in 3 days"
    Message="Renew your subscription to keep all functionality" />
```
![A mockup of an InfoBar in a Warning state with a close button and a message](images/Warning_TitleMessage.png)

## Programmatic dismiss in InfoBar
An InfoBar can be dismissed by the user via the close button or programmatically. 
If the notification is required to be in view until the status is resolved and you would like to remove 

> the ability for the user to dismiss the info bar, you can set the IsUserDismissable property to false.

Reviewer 2: "Close" is a more standard term for this action than "dismiss". And this should follow the "Can" pattern for consistency. So "CanClose" rather than "IsUserDismissable". (For example similar to [TextBox.CanUndo](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.TextBox.CanUndo).)

    Background="{StaticResource LavenderBackgroundBrush}">
    <controls:InfoBar.IconSource>
        <controls:SymbolIconSource Symbol="Phone" />
    </controls:InfoBar.IconSource>
</controls:InfoBar>
```

![A mockup of an InfoBar in the default state with a custom background color and icon](images/Custom_IconColor.png)

## Add an action button

An additional action button can be added by defining your own button that inherits 
[ButtonBase](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Controls.Primitives.ButtonBase) 
and setting it in the ActionButton property. Custom styling will be applied to action buttons of type 
[Button](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.Button) 
and [HyperlinkButton](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.HyperlinkButton) 
for consistency and accessibility. Outside of the ActionButton property, 
additional action buttons can be added via custom content and will appear below the message.

XAML
```xml
<controls:InfoBar x:Name="NoInternetNotification"
    Severity="Error"
    Title="No Internet"
    Message="Reconnect to continue working.">
    <controls:InfoBar.ActionButton>

>         <Button Content="Network Settings"/>

Reviewer 6: Do I need to add a `Click` handler to this `Button`, or is there an `ActionButtonClicked` event on the `InfoBar`?

Reviewer: You need to add the Click handler. There is no event for this on the InfoBar. Note that the action button is optional.

Reviewer 6: Suggest adding it to the sample then to make it clear :)

    </controls:InfoBar.ActionButton>
</controls:InfoBar>
```

![A mockup of an InfoBar in the Error state with a single line message and an action button](images/Error_ActionButton.png)

XAML
```xml
<controls:InfoBar x:Name="TermsAndConditionsUpdatedNotification"
    Title="Terms and Conditions Updated"
    Message="Dismissal of this message implies agreement to the updated Terms and Conditions. Please view the changes on our website.">
    <controls:InfoBar.ActionButton>
        <HyperlinkButton
            Content="Terms and Conditions Sep 2020" 

>             NavigateUri="http://www.microsoft.com"/>

Reviewer 6: Nit: should we use something like `example.com` rather than Microsoft?

    <controls:InfoBar.ActionButton/>
</controls:InfoBar>
```
![A mockup of an InfoBar with a message expanding multiple lines and a hyperlink](images/Default_Hyperlink.png)

## Content wrapping
The text set in the Message property will automatically wrap vertically in the control 
underneath the other visual components if the control is no longer able to expand horizontally. 

> The action button will also wrap in its own line.

Reviewer 2: Make sense. But is it really a paragraph or a new line?

XAML
```xml
<controls:InfoBar x:Name="BackupCompleteNotification"
    Severity="Success"
    Title="Backup complete: Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."  
    Message="All documents were uploaded successfully. Ultrices sagittis orci a scelerisque. Aliquet risus feugiat in ante metus dictum at tempor commodo. Auctor augue mauris augue neque gravida." />
```

![A mockup of an InfoBar in the Success state multiline title and message InfoBar](images/Success_ContentWrapping.png)

## Custom content
XAML content can be added to an InfoBar using the Content property. 

> It will appear in its own line below the rest of the control content. 

Reviewer 6: Below the `ActionButton` as well?

Author: Yep, see below:
![image](https://user-images.githubusercontent.com/20007011/93113016-81563480-f66d-11ea-937b-ec012d1077c6.png)

The InfoBar will expand to fit the content defined.

XAML
```xml
<controls:InfoBar x:Name="BackupInProgressNotification"
    Title="Backup in progress"  
    Message="Your documents are being saved to the cloud"
    IsUserDismissable="False">
    <controls:InfoBar.Content>
        <ProgressBar IsIndeterminate="True" Margin="0,0,0,6"/>
    </controls:InfoBar.Content>
</controls:InfoBar>
```

![A mockup of an InfoBar in its default state with an indeterminate progress bar](images/Default_CustomContent.gif)

## Lightweight styling

You can modify the default Style and ControlTemplate to give the control a unique appearance. 
For more info, see the 
[Light-weight styling section](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/xaml-styles#lightweight-styling)
of the 
[Styling controls](https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/xaml-styles)
article.

For example, the following causes the title bar font size to be 22pt on a Page's InfoBars:

```xml
<Page.Resources>
    <x:Double x:Key="InfoBarTitleFontSize">22</x:Double>
</Page.Resources>
```

# Inputs and Accessibility

## UI Automation Patterns 

> InfoBar will use a Pane control pattern for inline notifications and will implement a custom "information" Landmark.

Reviewer 7: Does a closed InfoBar appear in the automation tree at all?

### Keyboard Navigation 
- No action is needed to invoke the InfoBar
- After navigating to the control via tabbing, 

> the user is automatically focused into the control once navigating to the first actionable item.

Reviewer 2: I would just leave out the "in that order" in that suggestion. (Trying to say as little as possible about the current visualization)

> the user is automatically focused into the control once navigating to the first actionable item.

Author: I'll follow up on the close button automation with accessibility.
@StephenLPeters, what happens in Teaching Tip today? When the close button is pressed what is announced? Is the verbage specific to TeachingTip or is it just "close button"?

- Once focused, tabbing will iterate through all actionable items in the control in order. 
When tab is pressed on the last element in the InfoBar, focus will continue throughout the page.

>   - A user can also use the left and right arrow keys to navigate between the available buttons

Author: I'll follow up on this with accessibility.

> - To close the InfoBar the action or close button needs to be pressed.

Author: @MikeHillberg, who do you recommend I follow up on this with? Removing action button close functionality from spec for now as implementation doesn't include it.

  - Note: Escape will not close the InfoBar and will instead bubble up the command to the parent components.

### Gamepad
- The actionable items in the control can be navigated to via spatial navigation.
- The 'A' button will interact with the item in focus, such as "press" the action or close buttons

> - The InfoBar can be closed via pressing the 'X' close button, pressing the action button, or the 'B' button which will return focus to the element previously in focus.

Reviewer 2: Is it odd that 'B' doesn't behave like Escape?

Please view the guidance for 
[Adjust layout and fonts, and support RTL](https://docs.microsoft.com/en-us/windows/uwp/design/globalizing/adjust-layout-and-fonts--and-support-rtl) 
for more information about text localization in your notification.

# API Notes

### Notable Properties  
| Name | Description |
|:-:|:--|
| IsOpen| Gets or sets a value that determines the visibility of the InfoBar. By default, is set to false. |
| Severity | Gets or sets a value that indicates the  color and icon to style the InfoBar. It will also define the assistive technology settings. |
| IsUserDismissable| Gets or sets a boolean that indicates whether the user will be able to dismiss the InfoBar.

### Events  
| Name | Description |
|:-:|:--|
| CloseButtonClick | Occurs after the close button has been clicked. |
| Closed | Occurs after the info bar is closed. |
| Closing |Occurs just before the info bar begins to close. |

# API Details

```c++
enum InfoBarCloseReason
{
    CloseButton = 0,
    Programmatic = 1,
};

enum InfoBarSeverity
{

>     Error = 0,

Reviewer 6: Usually the "default" value is set as the `0` value, and so `Default` should be `0`.

Reviewer 2: Is Default the default?

    Warning = 1,
    Success = 2,
    Default = 3,
}

runtimeclass InfoBarClosedEventArgs
{
    InfoBarCloseReason Reason{ get; };
};

runtimeclass InfoBarClosingEventArgs
{
    InfoBarCloseReason Reason{ get; };
    Boolean Cancel;

>     Windows.Foundation.Deferral GetDeferral();

Reviewer 6: Is there a good scenario for deferral? Is there a deadline for the deferral? Is it any different than the app specifying `Cancel = true`, disabling input on the control, doing the async work, then programmatically closing the bar?

Author: One example scenario is if the InfoBar is showing a "Terms and Conditions are updated and need to be accepted" message and there's an inline checkbox that a user needs to check before the InfoBar closes. If the user manually closes the InfoBar but the checkbox isn't checked then the close could be cancelled.

@teaP for the deferral deadline and replicating deferral behavior questions?

Reviewer 6: `Cancel` makes perfect sense; it's the deferral that is potentially not useful. Usually it is used if you need to make a network connection or write to a brokered file.

Reviewer 2: In this case it's to enable a dialog and user interaction before closing the InfoBar, which like files and network is async.

};

unsealed runtimeclass InfoBarTemplateSettings : Windows.UI.Xaml.DependencyObject
{
    InfoBarTemplateSettings();
    Windows.UI.Xaml.Controls.IconElement IconElement;

    static Windows.UI.Xaml.DependencyProperty IconElementProperty{ get; };
}

unsealed runtimeclass InfoBar : Windows.UI.Xaml.Controls.ContentControl
{
    InfoBar();

    String Title;
    String Message;

    Boolean IsOpen;

>     Boolean IsUserDismissable;

Reviewer 6: Should this be `IsCloseButtonVisible`?

Author: @mikeh brought this up earlier in the [conversation](https://github.com/microsoft/microsoft-ui-xaml-specs/pull/94#discussion_r487346017):
> "Close" is a more standard term for this action than "dismiss". And this should follow the "Can" pattern for consistency. So "CanClose" rather than "IsUserDismissable". (For example similar to [TextBox.CanUndo](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.TextBox.CanUndo).)

We avoided IsCloseButtonVisible to remain flexible with devices in the future (like touch-enabled) where a close button may not be visible but the user is still able to close the bar via a swipe or other gesture. What are your thoughts on CanClose?

Reviewer 8: Guessing you mis-tagged and this was meant for @MikeHillberg :-)

Reviewer 6: Sure, `CanClose` is good too.

    Boolean IsIconVisible;

    Windows.UI.Xaml.Controls.Primitives.ButtonBase ActionButton;

    Windows.UI.Xaml.Style CloseButtonStyle;
    Windows.UI.Xaml.Input.ICommand CloseButtonCommand;
    Object CloseButtonCommandParameter;

    InfoBarSeverity Severity;
    IconSource IconSource;

    InfoBarTemplateSettings TemplateSettings{ get; };
    Object Content{ get; set; };
    Windows.UI.Xaml.DataTemplate ContentTemplate{ get; set; };

    event Windows.Foundation.TypedEventHandler<InfoBar, Object> CloseButtonClick;

>     event Windows.Foundation.TypedEventHandler<InfoBar, InfoBarClosingEventArgs> Closing;

Reviewer 7: The fact that you can trigger `Closing` by setting `IsOpen=false` means that inside the `Closing` handler, if you read the `IsOpen` property, you get `false` which is weird because you expect `Closing` to be raised *before* it closes. (As noted later, if you then cancel the Closing event, this sets `IsOpen` back to `true`.)

If the InfoBar is interactively closed, does `IsOpen` read as true or false during `Closing`?

`NavigationView` has the same problem, since you can set `IsPaneOpen = !IsPaneOpen` and the `-ing` events will be raised even though the transition has already occurred as far as the property is concerned. Are we at least consistent with NavigationView's behavior here?

Reviewer 2: +1 to consistency, see also [FlyoutBase.IsOpen](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Controls.Primitives.FlyoutBase.IsOpen)

Reviewer 9: For NavigationView, we're toggling the IsPaneOpen property as soon as the action begins. So in both PaneClosing and PaneClosed events, we report IsPaneOpen as False; in both PaneOpening and PaneOpen we report IsPaneOpen as True.

Reviewer 9: Also, unless I'm doing something wrong, cancelling the event using NavigationViewPaneClosingEventArgs does nothing. The behavior of this
```c#
        private void NavView_PaneClosing(Microsoft.UI.Xaml.Controls.NavigationView sender, Microsoft.UI.Xaml.Controls.NavigationViewPaneClosingEventArgs args)
        {
            bool open = sender.IsPaneOpen;
            args.Cancel = true;
            open = sender.IsPaneOpen;
        }
```
is identical to the behavior of this
```c#
        private void NavView_PaneClosing(Microsoft.UI.Xaml.Controls.NavigationView sender, Microsoft.UI.Xaml.Controls.NavigationViewPaneClosingEventArgs args)
        {
            bool open = sender.IsPaneOpen;
        }
```

Reviewer 10: @YuliKI We have the open issue https://github.com/microsoft/microsoft-ui-xaml/issues/3352 which I plan to look into soon and where I could take a look at IsPaneOpen not being updated correctly when cancelling the close.

Reviewer 9: Thank you, @Felix-Dev, that would be great.

Author: Wrapping up this comment thread, yes, the behavior for InfoBar is the same as NavView. Once closing or opening starts, the isOpen property is set to false or true respectively.

Author: Thanks @YuliKl and @Felix-Dev for hopping in here. InfoBar and NavigationView _do_ follow the same pattern here that immediately when the action begins IsOpen is set to the intended value. See RaiseClosingEvent and RaiseClosedEvent in [InfoBar.cpp](https://github.com/microsoft/microsoft-ui-xaml/pull/3325/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.cs&file-filters%5B%5D=.h&file-filters%5B%5D=.projitems&file-filters%5B%5D=.shproj&file-filters%5B%5D=.targets&file-filters%5B%5D=.vcxproj&file-filters%5B%5D=.xaml#diff-624e3a295bc0485e93ffbc0dc6e381a4240ab7fc621f7600f3ad401ef3942963)

Author: @teaP ping to ensure I'm correct here 😊

    static Windows.UI.Xaml.DependencyProperty ActionButton{ get; };
    static Windows.UI.Xaml.DependencyProperty ContentProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty ContentTemplateProperty{ get; };
    static Windows.UI.Xaml.DependencyProperty TemplateSettingsProperty{ get; };
}
```

## Theme Resources

### Notable Theme Resources
| Name | Description |
|:-:|:--|
|InfoBarHyperlinkForeground | Sets the hyperlink button text color. <br> - Note: This is set to keep hyperlinks accessible on the variously colored backgrounds defined by the severity background colors.
|InfoBarCloseButtonStyle | Sets the close button style of the InfoBar and contains the following properties: <br> - Width <br> - Height <br> - VerticalAlignment <br> - Background <br> - Margin <br> - CornerRadius
|InfoBarIconFontSize | Sets the icon size of the InfoBar. <br> - Note: Custom icons set via the IconSource property will not resize to 20px unless they are in the standard symbol set.

### All Theme Resources
| Resource Key | Description |
|:-:|:--|
| InfoBarSeverityErrorBackgroundBrush | Error Severity background color 

> | InfoBarSeverityWarningBackgroundBrush | Warning Severity background color 

Reviewer 2: As the description implies, this should be ...**WarningSeverity**BackgroundBrush rather than ...SeverityWarning...

| InfoBarIconFontSize | Icon font size
| InfoBarTitleHorizontalLayoutMargin | Horizontal, single-line, layout title margin thickness
| InfoBarTitleVerticalLayoutMargin | Vertical, multiline, layout title margin thickness
| InfoBarMessageHorizontalLayoutMargin | Horizontal, single-line, layout message margin thickness
| InfoBarMessageVerticalLayoutMargin | Vertical, multiline, layout message margin thickness
| InfoBarActionHorizontalLayoutMargin | Horizontal, single-line, layout action button margin thickness
| InfoBarActionVerticalLayoutMargin | Vertical, multiline, layout action button margin thickness
| InfoBarActionButtonMinWidth | Action button minimum width
| InfoBarActionButtonHeight | Action button height
| InfoBarActionButtonPadding | Action button padding thickness
| InfoBarActionButtonCornerRadius | Action button corner radius
| InfoBarHyperlinkButtonFontSize | Hyperlink button font size
| InfoBarHyperlinkButtonHeight | Hyperlink button height
| InfoBarHyperlinkButtonPadding | Hyperlink button padding thickness
| InfoBarCloseButtonSymbol | Close button symbol glyph
| InfoBarCloseButtonStyle | Close button style

# Appendix

## Design References
UI Elements for InfoBar

> ![Toolkit page for InfoBar UI elements](images/Docked_Toolkit.png)

Author: Need to update this with design.

| Icon | - Defined by either the Severity or by IconSource <br> - Can be removed by setting IsIconVisible to false
| Title | - Semi-bolded and appears right of the Icon <br> - Recommended to be 50 characters or less
| Message | - Will appear to the right of the Title in single-height notifications, otherwise will be on a new line <br> - Recommended to be 512 characters or less
| Action button |  - Will appear directly to the right of the Message in single height notifications, otherwise will be on its own new line. <br> - An action button can be any class that inherits ButtonBase however Button, HyperlinkButton, and Checkbox are our recommended options <br> - Additional action buttons may be added through custom XAML content
| Close button | - Will appear as 'X' by default <br> - Can be removed by setting IsUserDismissable to false
| Content | - Can be customizable to include text, hyperlinks, and any other XAML content <br> - Appears in its own line underneath the other InfoBar content

## Behavioral Components
| Property | Notes |
|:---:|:---|
| Opening | - An info bar is shown by setting its IsOpen property to true. |

> | Closing | There are two ways an info bar can close: <br>- The program sets the IsOpen property to false <br> - The user invokes the Close button. <br> Use the InfoBarCloseReason to determine which case has occurred. <br> Closing can be prevented by setting the Cancel property to true. You can use a deferral to respond asynchronously to the event. |

Reviewer 7: 3. The user presses the gamepad B button.

### Canceling and deferring close
The Closing event can be used to cancel and/or defer the close of an InfoBar. 
This can be used to keep the InfoBar open or allow time for a custom action to occur. 
When the closing of an InfoBar is canceled, IsOpen will go back to true, however, 
it will stay false during the deferral. A programmatic close can also be canceled.

XAML
```xml
<controls:InfoBar x:Name="UpdateAvailable"
    Title="Update Available"
    Message="Please close this tip to apply required security updates to this application"
    Closing="InfoBar_Closing">
</controls:InfoBar>
```
C#
```C#
public void InfoBar_Closing(InfoBar sender, InfoBarClosingEventArgs args)
{
    if (args.Reason == InfoBarCloseReason.CloseButton) {

>         // if scenario failed

Reviewer 7: Use of comment here is confusing. This isn't commenting the code; it's representing code that is missing. Better would be to call a function not implemented here.

```c#
if (args.Reason == InfoBarCloseReason.CloseButton) {
  if (!ApplyUpdates()) {
    // could not apply updates - do not close
    args.Cancel = true;
  }
}
```
